### PR TITLE
prepare lib refactoring improve tests

### PR DESCRIPTION
**Pull request by maxy-shpfy**
## Description

This PR adds comprehensive test coverage for the component library and flow canvas functionality. It includes new unit tests for the ComponentLibraryProvider, task node inputs validation, and improves existing tests by using React Testing Library's `act` function to properly handle asynchronous rendering. The PR also adds data-testid attributes to various components to facilitate testing and adds end-to-end tests for component library interactions.

## Type of Change

- [x] Improvement
- [x] Cleanup/Refactor

## Checklist

- [x] I have tested this does not break current pipelines / runs functionality
- [ ] I have tested the changes on staging

## Test Instructions

1. Run the unit tests to verify the new test coverage for ComponentLibraryProvider and task node inputs
2. Run the e2e tests to verify component library interactions work as expected
3. Verify that the RunDetails tests still pass with the updated testing approach

## Additional Comments

The PR adds two new e2e test files:

- `componentlib.spec.ts` - Tests component library functionality including searching, favoriting, and dragging components
- `ComponentLibraryProvider.test.tsx` - Unit tests for the component library provider

It also adds data-testid attributes to various components to make them more testable, particularly for the node handles and connections in the flow canvas.

### CODE REVIEW

**Comment by maxy-shpfy** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/714?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#719** <a href="https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/719?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#718** <a href="https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/718?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#717** <a href="https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/717?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#716** <a href="https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/716?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#715** <a href="https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/715?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#714** <a href="https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/714?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/714?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#713** <a href="https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/713?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#645** <a href="https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/645?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#595** <a href="https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/595?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#596](https://github.com/Cloud-Pipelines/pipeline-studio-app/pull/596) <a href="https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/596?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/components/PipelineRun/RunDetails.test.tsx [hunk 1]
```diff
@@ -174,7 +176,7 @@ describe("<RunDetails/>", () => {
       });
 
       // act
-      renderWithQueryClient(<RunDetails />);
+      await act(async () => renderWithQueryClient(<RunDetails />));
```
Comment (maxy-shpfy) [inline src/components/PipelineRun/RunDetails.test.tsx:179]: Here and in all similar cases - wrapping with `act` prevents vitest from throwing warnings

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts [hunk 2]
```diff
@@ -0,0 +1,669 @@
...
+      expect(result.folders?.[0]?.components?.[0]?.spec?.name).toBe(
+        "parsed-parent-yaml",
+      );
+      expect(
+        result.folders?.[0]?.folders?.[0]?.components?.[0]?.spec?.name,
+      ).toBe("parsed-child-yaml");
+    });
+
+    // todo: test error handling when parsing fails
+    // todo: test when URL fetch fails
```
Comment (camielvs) [inline src/providers/ComponentLibraryProvider/componentLibrary.test.ts:498]: todo now, or later?

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts [hunk 3]
```diff
@@ -0,0 +1,669 @@
...
+      // Act
+      const result = flattenFolders(folder);
+
+      // Assert
+      expect(result).toHaveLength(2);
+      expect(result[0].name).toBe("component-1");
+      expect(result[1].name).toBe("component-2");
+    });
+
+    // todo: test with malformed folder structures
```
Comment (camielvs) [inline src/providers/ComponentLibraryProvider/componentLibrary.test.ts:583]: todo now or later?

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+
+    it("should handle empty array", () => {
+      // Act
+      const result = filterToUniqueByDigest([]);
+
+      // Assert
+      expect(result).toEqual([]);
+    });
+
+    // todo: test with null/undefined digest values
```
Comment (camielvs) [inline src/providers/ComponentLibraryProvider/componentLibrary.test.ts:643]: todo now or later?

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+    it("should work without user components", () => {
+      // Act
+      const result = fetchFavoriteComponents(mockComponentLibrary, undefined);
+
+      // Assert
+      expect(result.components).toHaveLength(2);
+      expect(result.components?.every((c) => c.favorited)).toBe(true);
+    });
+
+    // todo: test deduplication of favorites with same digest across library and user components
```
Comment (camielvs) [inline src/providers/ComponentLibraryProvider/componentLibrary.test.ts:365]: todo now or later?

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts [hunk 6]
```diff
@@ -0,0 +1,669 @@
...
+        folders: [],
+        isUserFolder: false,
+      };
+      expect(result1).toEqual(expectedResult);
+      expect(result2).toEqual(expectedResult);
+      expect(result3).toEqual(expectedResult);
+    });
+
+    // todo: test with components without digest
+    // todo: test with deeply nested task structures
```
Comment (camielvs) [inline src/providers/ComponentLibraryProvider/componentLibrary.test.ts:257]: todo now or later?

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts [hunk 7]
```diff
@@ -0,0 +1,669 @@
...
+      });
+      expect(consoleSpy).toHaveBeenCalledWith(
+        "Error fetching user components:",
+        expect.any(Error),
+      );
+
+      consoleSpy.mockRestore();
+    });
+
+    // todo: test edge cases with malformed component files
```
Comment (camielvs) [inline src/providers/ComponentLibraryProvider/componentLibrary.test.ts:161]: todo now or later?

File: src/providers/ComponentLibraryProvider/index.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export {
+  ComponentLibraryProvider,
+  useComponentLibrary,
+} from "./ComponentLibraryProvider";
```
Comment (camielvs) [inline src/providers/ComponentLibraryProvider/index.ts:4]: I may be wrong but did we agree to avoid barrel files and index exports?

File: tests/e2e/componentlib.spec.ts [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+
+    // expect the component to be in the favorites folder
+    const favoritesFolder = await openComponentLibFolder(page, "My Components");
+    expect(await favoritesFolder.locator("li")).toHaveCount(1);
+
+    // unstar the component
+    await chicagoTaxiTripsDataset.getByTestId("favorite-star").click();
+
+    // giving time for the component to be removed from the favorites folder
+    // todo: find a better way to do this
```
Comment (camielvs) [inline tests/e2e/componentlib.spec.ts:131]: todo now or later?

File: src/providers/ComponentLibraryProvider/index.ts (hunk 8 above)
Comment (maxy-shpfy) [reply src/providers/ComponentLibraryProvider/index.ts:4]: Yes, indeed. In this case, while refactoring is in progress, I decided to use barrel files, to limit the blast impact to the rest of the app.

File: tests/e2e/componentlib.spec.ts (hunk 9 above)
Comment (maxy-shpfy) [reply tests/e2e/componentlib.spec.ts:131]: All todos are for later.

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts (hunk 2 above)
Comment (maxy-shpfy) [reply src/providers/ComponentLibraryProvider/componentLibrary.test.ts:498]: All todos are for later. It may depends on the deepness of changes.

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts (hunk 3 above)
Comment (maxy-shpfy) [reply src/providers/ComponentLibraryProvider/componentLibrary.test.ts:583]: All todos are for later. It may depends on the deepness of changes.

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts (hunk 4 above)
Comment (maxy-shpfy) [reply src/providers/ComponentLibraryProvider/componentLibrary.test.ts:643]: All todos are for later. It may depends on the deepness of changes.

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts (hunk 5 above)
Comment (maxy-shpfy) [reply src/providers/ComponentLibraryProvider/componentLibrary.test.ts:365]: All todos are for later. It may depends on the deepness of changes.

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts (hunk 6 above)
Comment (maxy-shpfy) [reply src/providers/ComponentLibraryProvider/componentLibrary.test.ts:257]: All todos are for later. It may depends on the deepness of changes.

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts (hunk 7 above)
Comment (maxy-shpfy) [reply src/providers/ComponentLibraryProvider/componentLibrary.test.ts:161]: All todos are for later. It may depends on the deepness of changes.

File: src/components/shared/ReactFlow/FlowCanvas/TaskNode/TaskNodeCard/TaskNodeInputs.test.tsx [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+    defaultValue?: string,
+  ): InputSpec => ({
+    name,
+    type,
+    optional,
+    default: defaultValue,
+  });
+
+  const createMockTaskSpec = (
+    taskArguments?: Record<string, any>,
```
Comment (camielvs) [inline src/components/shared/ReactFlow/FlowCanvas/TaskNode/TaskNodeCard/TaskNodeInputs.test.tsx]:
```suggestion
    taskArguments?: Record<string, ArgumentType>,
```

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts [hunk 11]
```diff
@@ -0,0 +1,669 @@
...
+          {
+            name: "Test Component 1",
+            creationTime: new Date(),
+            modificationTime: new Date(),
+            data: new ArrayBuffer(0),
+            componentRef: {
+              name: "test-component-1",
+              digest: "digest1",
+              url: "https://example.com/component1.yaml",
+              spec: {} as any,
```
Comment (camielvs) [inline src/providers/ComponentLibraryProvider/componentLibrary.test.ts]: as `ComponentSpec`?

File: src/providers/ComponentLibraryProvider/componentLibrary.test.ts [hunk 12]
```diff
@@ -0,0 +1,669 @@
...
+      const result = fetchUsedComponents(mockGraphSpec);
+
+      // Assert
+      expect(result.components).toHaveLength(1);
+      expect(result.components?.[0]?.digest).toBe("same-digest");
+    });
+
+    it("should return empty folder when graphSpec is invalid", () => {
+      // Act
+      const result1 = fetchUsedComponents(null as any);
```
Comment (camielvs) [inline src/providers/ComponentLibraryProvider/componentLibrary.test.ts:240]: `null as any`? this seems weird. surely our typeguarding would take care of the `null` case

**Comment by maxy-shpfy** [comment]
### Merge activity

* **Aug 28, 11:11 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/Cloud-Pipelines/pipeline-studio-app/714).
